File: codes/liskov/liskhely/liskovsertve.cpp
#include <iostream>

class Bird
{
    
};

class IFlyingBird : public Bird
{
public:
    virtual void fly() = 0;
};
class INotFlyingBird : public Bird
{
};

class Eagle : public IFlyingBird
{
public:
  
    void fly() override
    {
        std::cout << "Eagle: flying..\n";
    }
};

class Penguin : public INotFlyingBird
{
};

static void flyBird(IFlyingBird& b)
{
    b.fly();
}

int main()
{
    Eagle theEagle;
    Penguin thePenguin;
    
    flyBird(theEagle);
    //flyBird(thePenguin); //Ez az utasításmár nem fog lefordulni...
    //A programtulajdonságai épek maradnak.
    
    return 0;
}

File: codes/liskov/liskhely/liskovsertve.java
interface Bird
{
    
}

interface IFlyingBird extends Bird
{
    public void  fly();
}
interface INotFlyingBird extends Bird
{
}
class Eagle implements IFlyingBird
{
    public void fly()
    {
        System.out.println("Eagle: flying..\n");
    }
}
class Penguin implements INotFlyingBird
{
}

class LiskovSub
{

    public static void flyBird(IFlyingBird b)
    {
        b.fly();
    }

    public static void main(String[] args)
    {
        Eagle theEagle = new Eagle();
        Penguin  thePenguin = new Penguin();
        
        flyBird(theEagle);
        //flyBird(thePenguin); //Fordítási hiba!!
        
        
    }

} 

